lang/compiler.go

```diff
@@ -337,3 +340,127 @@ func (p *AstProgram) Start(ctx context.Context, alloc *memory.Allocator) (flux.Q
 
 	return p.Program.Start(ctx, alloc)
 }
+
+func (p *AstProgram) eval(ctx context.Context, deps dependencies.Interface) ([]interpreter.SideEffect, values.Scope, error) {
+	s, _ := opentracing.StartSpanFromContext(ctx, "eval")
+
+	sideEffects, scope, err := flux.EvalAST(ctx, deps, p.Ast, flux.SetNowOption(p.Now))
+	if err != nil {
+		return nil, nil, err
+	}
+
+	s.Finish()
+	s, _ = opentracing.StartSpanFromContext(ctx, "compile")
+	defer s.Finish()
+
+	nowOpt, ok := scope.Lookup(flux.NowOption)
+	if !ok {
+		return nil, nil, fmt.Errorf("%q option not set", flux.NowOption)
+	}
+
+	nowTime, err := nowOpt.Function().Call(ctx, deps, nil)
+	if err != nil {
+		return nil, nil, err
+	}
+	p.Now = nowTime.Time().Time()
+	return sideEffects, scope, nil
+}
+
+func (p *AstProgram) updateOpts(scope values.Scope) error {
+	planOpts, err := getPlanOptions(scope)
+	if err != nil {
+		return err
+	}
+	p.opts.planOptions.logical = append(p.opts.planOptions.logical, planOpts.logical...)
+	p.opts.planOptions.physical = append(p.opts.planOptions.physical, planOpts.physical...)
+	return nil
+}
+
+func getPlanOptions(scope values.Scope) (planOptions, error) {
+	planOpt, ok := scope.Lookup("planner")
+	if !ok {
+		// No option specified.
+		return planOptions{}, nil
+	}
+	if planOpt.Type().Nature() != semantic.Object {
+		return planOptions{}, fmt.Errorf("option 'planner' must be an object, got %s", planOpt.Type().Nature().String())
+	}
+	var err error
+	var wrongKeysErr = fmt.Errorf("the only available field for option 'planner' is 'disable'")
+	if planOpt.Object().Len() > 3 {
+		return planOptions{}, wrongKeysErr
+	}
+	planOpt.Object().Range(func(name string, v values.Value) {
+		// TODO(affo): we could add 'enable' and 'only'.
+		//  In order to do this we should add the possibility to enable rules by name in the planner, first.
+		if name != "disable" {
+			err = wrongKeysErr
+		}
+	})
+	if err != nil {
+		return planOptions{}, err
+	}
+
+	la, pa, err := getRuleArraysForKey("disable", planOpt)
+	if err != nil {
+		return planOptions{}, err
+	}
+
+	ls := make([]plan.LogicalOption, la.Array().Len())
+	ps := make([]plan.PhysicalOption, pa.Array().Len())
+	la.Array().Range(func(i int, v values.Value) {
+		ls[i] = plan.RemoveLogicalRule(v.Str())
+	})
+	pa.Array().Range(func(i int, v values.Value) {
+		ps[i] = plan.RemovePhysicalRule(v.Str())
+	})
+	return planOptions{
+		logical:  ls,
+		physical: ps,
+	}, nil
+}
+
+func getRuleArraysForKey(key string, planOpt values.Value) (values.Array, values.Array, error) {
+	lv := values.NewArray(semantic.String)
```

Comment:
Is it for naming? See above.

In this case I see the possibility of refactoring to `logicalValues`. This is because they are used throughout the function.